infrastructure/src/main/java/org/corfudb/infrastructure/management/PollReport.java

```diff
@@ -21,43 +23,78 @@
-     * to failedNodes list.
+     * Current cluster state, collected by a failure detector
      */
-    @Default
-    private final ImmutableSet<String> failedNodes = ImmutableSet.of();
+    @NonNull
+    private final ClusterState clusterState;
+
     /**
-     * A map of nodes answered with {@link WrongEpochException}.
+     * List of responsive servers in a current layout
      */
-    @Default
-    private final ImmutableMap<String, Long> wrongEpochs = ImmutableMap.of();
+    @NonNull
+    private final List<String> activeLayoutServers;
+
     /**
-     * Indicates if current layout slot is unfilled
+     * Returns all connected nodes to the current node including nodes with higher epoch.
+     *
+     * @return set of all reachable nodes
      */
-    @Default
-    private final boolean currentLayoutSlotUnFilled = false;
+    public ImmutableSet<String> getAllReachableNodes() {
+        return Sets.union(getConnectedNodes(), wrongEpochs.keySet()).immutableCopy();
+    }
+
     /**
-     * Current cluster state, collected by a failure detector
+     * Contains list of servers successfully connected with current node.
+     * It doesn't contain nodes answered with WrongEpochException.
      */
-    @NonNull
-    private final ClusterState clusterState;
+    public Set<String> getConnectedNodes() {
+        return clusterState
+                .getLocalNodeConnectivity()
+                .getConnectedNodes();
+    }
 
     /**
-     * Returns all connected nodes to the current node
-     * @return
+     * Contains list of servers disconnected from this node. If the node A can't ping node B then node B will be added
+     * to failedNodes list.
      */
-    public ImmutableSet<String> getAllConnectedNodes() {
-        return Sets.union(connectedNodes, wrongEpochs.keySet()).immutableCopy();
+    public Set<String> getFailedNodes() {
+        Set<String> failedNodes = clusterState.getLocalNodeConnectivity().getFailedNodes();
+
+        return Sets.difference(failedNodes, wrongEpochs.keySet());
+    }
+
+    /**
+     * All active Layout servers have been sealed but there is no client to take this forward and
+     * fill the slot by proposing a new layout. This is determined by the outOfPhaseEpochNodes map.
+     *
+     * @return True if latest layout slot is vacant.
+     */
+    public boolean isCurrentLayoutSlotUnFilled() {
+        log.trace("isCurrentLayoutSlotUnFilled. Wrong epochs: {}, active servers: {}",
+                wrongEpochs, activeLayoutServers
+        );
+
+        // Check if all active layout servers are present in the outOfPhaseEpochNodes map.
+        boolean result = wrongEpochs.keySet().containsAll(activeLayoutServers);
+
+        if (result) {
+            String msg = "Current layout slot is empty. Filling slot with current layout." +
+                    " wrong epochs: {}, active layout servers: {}";
+            log.info(msg, wrongEpochs, activeLayoutServers);
```

Comment:
Can you put the log message directly here?

Reply:
done